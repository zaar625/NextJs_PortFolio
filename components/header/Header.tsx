'use client'

import React,{useState,useEffect,useRef} from 'react';
import { usePathname } from 'next/navigation'
import Link from 'next/link';
import {AiOutlineMenu,AiOutlineClose} from 'react-icons/ai'
import { headerNav , productNav} from '@/constant/navigation';
import { getAuth } from "firebase/auth";
import {app} from '@/lib/firebaseConfig';
import { useSession,signOut } from "next-auth/react"

import './header.scss';


export default function Header() {
  const [menuIsActive, setMenuIsActive] = useState(false) 
  const { data: snsSession } = useSession()
  
  const headerRef = useRef<HTMLDivElement>(null);
  const pathName = usePathname();
  const auth = getAuth(app);
  const firebaseUser = auth.currentUser 

  useEffect(() => {
    const shrinkHeader = () => {
      if (headerRef !== null) {
        if (
          document.body.scrollTop > 100 ||
          document.documentElement.scrollTop > 100
        ) {
          headerRef.current?.classList.add('shrink');
        } else {
          headerRef.current?.classList.remove('shrink');
        }
      }
    };
    window.addEventListener('scroll', shrinkHeader);
    return () => {
      window.removeEventListener('scroll', shrinkHeader);
    };
  }, []);

  const logout = () => {
    if (window.confirm('로그아웃하시겠습니까?')) {
      if(firebaseUser){
        auth.signOut();
      }

      if(snsSession){
        signOut();
      }

      alert('로그아웃되었습니다.');
      window.location.replace('/');
    }
  }

  return (
    <header ref ={headerRef}className='header'>
      <nav className='header__wrap'>
        <div className='header__top container'>
          <p className='logo'>
            <Link href={'/'}>
              BABAN
            </Link>
          </p>
          <button className='menu' onClick={()=>setMenuIsActive(!menuIsActive) }>
          {menuIsActive ? <AiOutlineClose/> : <AiOutlineMenu/> }   
          </button>
          <ul className='header__nav-right'>
          <li>
            {firebaseUser || snsSession ? <button onClick={logout}>logout</button> : <Link href={'/login'}>login</Link>}
          </li>
          {headerNav.map((link, index) => {
          const isActive = pathName === link.path;
          return (
            <li key={index}>
              <Link  href={link.path} key={link.display} className={`${isActive ? 'active' : ''}`}>{link.display}</Link>
            </li>
          )
        })}
          </ul>
        </div>
          {menuIsActive &&(
            <ul className='product-nav'>
              {productNav.map((link, index) => {
              return (
                <li key={index}>
                  <Link href={`/${link.display}`} key={link.display}>{link.display}</Link>
                </li>
              )
            })}
            </ul>) 
          }
      </nav>
     
    </header>
  )
}
