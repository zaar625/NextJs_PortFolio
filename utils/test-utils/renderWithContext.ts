import { render } from "@testing-library/react";
import { ReduxProvider } from "@/redux/ReduxProvider";

const renderWithContext = (ui:React.JSX.Element) =>
  render(ui, { wrapper: ReduxProvider});

// re-export everything
export * from "@testing-library/react";

// override render method
export { renderWithContext as render };