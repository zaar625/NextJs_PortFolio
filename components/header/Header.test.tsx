import { render, screen } from '@/utils/test-utils/renderWithContext';
import Header from './Header';

jest.mock('next-auth/react', () => {
  const originalModule = jest.requireActual('next-auth/react');
  const mockSession = {
    expires: new Date(Date.now() + 2 * 86400).toISOString(),
    user: { username: 'tester' }
  };
  return {
    __esModule: true,
    ...originalModule,
    useSession: jest.fn(() => {
      return { data: mockSession, status: 'authenticated' }; // return type is [] in v3 but changed to {} in v4
    })
  };
});

test('헤더에 로고가 있는지 확인', async () => {
  render(<Header />);
  const logoText = await screen.findByRole('link', { name: 'BABAN' });

  expect(logoText).toBeInTheDocument();
});

describe('사용자의 로그인 상태 여부 확인', () => {
  it('renders login button when user is not authenticated', () => {
    // 모킹된 useSession 함수의 반환값을 설정합니다.
    const mockUseSession = jest.requireMock('next-auth/react').useSession;
    mockUseSession.mockReturnValueOnce({ data: null });

    // Header 컴포넌트를 렌더링합니다.
    render(<Header />);

    // 'login' 텍스트가 렌더링되었는지 확인합니다.
    expect(screen.getByText('login')).toBeInTheDocument();
  });

  it('renders logout button when user is authenticated', () => {
    // 모킹된 useSession 함수의 반환값을 설정합니다.
    const mockUseSession = jest.requireMock('next-auth/react').useSession;
    mockUseSession.mockReturnValueOnce({ data: { user: { name: 'testUser' } } });

    // Header 컴포넌트를 렌더링합니다.
    render(<Header />);

    //'logout' 텍스트가 렌더링되었는지 확인합니다.
    expect(screen.getByText('logout')).toBeInTheDocument();
  });
});
