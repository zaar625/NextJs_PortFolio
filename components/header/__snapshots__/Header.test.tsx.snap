// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`사용자가 로그인상태라면 로그아웃 링크가 있어야합니다. 1`] = `Promise {}`;
